src/capture/capture_video_ffmpeg.cpp

```diff
@@ -0,0 +1,733 @@
+		return AV_PIX_FMT_BGR565;
+	case PixelFormat::BGR888:
+		return AV_PIX_FMT_BGR24;
+	case PixelFormat::BGRX8888:
+		return AV_PIX_FMT_BGR0;
+	default:
+		assertm(false, "Invalid PixelFormat value");
+		return AV_PIX_FMT_NONE;
+	}
+}
+
+void FfmpegEncoder::CaptureVideoAddFrame(const RenderedImage& image,
+                                            const float frames_per_second)
+{
+	const int rounded_fps = iroundf(frames_per_second);
+	if (!video_encoder.is_initalised) {
+		if (!video_encoder.Init(image.params.width,
+		                        image.params.height,
+		                        image.params.pixel_format,
+		                        rounded_fps)) {
+			LOG_ERR("FFMPEG: Video encoder failed to init");
+			video_encoder.Free();
+			return;
+		}
+
+		bool init_muxer = false;
+		if (!muxer.is_initalised && audio_encoder.is_initalised) {
+			init_muxer = muxer.Init(video_encoder, audio_encoder);
+			if (!init_muxer) {
+				LOG_ERR("FFMPEG: Failed to init muxer");
+				muxer.Free();
+				video_encoder.Free();
+				return;
+			}
+		}
+
+		mutex.lock();
+		video_encoder.is_initalised = true;
+		muxer.is_initalised         = muxer.is_initalised || init_muxer;
+		mutex.unlock();
+		waiter.notify_all();
+
+	} else if (video_encoder.width != image.params.width ||
+	           video_encoder.height != image.params.height ||
+	           video_encoder.pixel_format != image.params.pixel_format ||
+	           video_encoder.frames_per_second != rounded_fps) {
+
+		CaptureVideoFinalise();
+		if (!video_encoder.Init(image.params.width,
+		                        image.params.height,
+		                        image.params.pixel_format,
+		                        rounded_fps)) {
+			LOG_ERR("FFMPEG: Video encoder failed to init");
+			video_encoder.Free();
+			return;
+		}
+
+		mutex.lock();
+		video_encoder.is_initalised = true;
+		mutex.unlock();
+		waiter.notify_all();
+	}
+
+	RenderedImage copy = image.deep_copy();
+	if (!video_encoder.queue.Enqueue(std::move(copy))) {
+		// Should never happen but avoid a memory leak if it does
+		LOG_ERR("FFMPEG: Video encoder queue is stopped");
+		copy.free();
+	}
+}
+
+void FfmpegEncoder::CaptureVideoAddAudioData(const uint32_t sample_rate,
+                                                 const uint32_t num_sample_frames,
+                                                 const int16_t* sample_frames)
+{
+	if (!audio_encoder.is_initalised) {
+		if (!audio_encoder.Init(sample_rate)) {
+			LOG_ERR("FFMPEG: Failed to init audio encoder");
+			audio_encoder.Free();
+			return;
+		}
+
+		bool init_muxer = false;
+		if (!muxer.is_initalised && video_encoder.is_initalised) {
+			init_muxer = muxer.Init(video_encoder, audio_encoder);
+			if (!init_muxer) {
+				LOG_ERR("FFMPEG: Failed to init muxer");
+				muxer.Free();
+				audio_encoder.Free();
+				return;
+			}
+		}
+
+		mutex.lock();
+		audio_encoder.is_initalised = true;
+		muxer.is_initalised         = muxer.is_initalised || init_muxer;
```

Comment:
Blank line above this perhaps so the auto-formatting looks nicer.

Reply:
fixed